### PR TITLE
Multi-Actions: Add support for Mistral AI

### LINKED ISSUE #5169: [Multi Actions] Make it work with Mistral models
Missing support for:
- `function_calls` in msg history
- `functions` in msg history

### PR DESCRIPTION
**Pull request by spolu**
## Description

Fixes: https://github.com/dust-tt/dust/issues/5169

- Adds support for tool_call_id
- Adds support for multiple function calls
- Make sure we can interact with 8x22b in Dust apps
- Fix our stitching of assistant message before tool message in the case of multiple tool messages

Tested with `mistral-small-latest`, `mistral-large-latest`  and `open-mixtral-8x22b`

Mistral wants as many `tool` messages (`function` for us) as there were `tool_calls`, `function_calls` for us).

`mistral-small-latest` does not seem to be able to do multiple tool calls at the same time but the two other do. 

Screenshot from 2024-05-29 12-12-08
Screenshot from 2024-05-29 12-11-47

TODO:
- [x] test with streaming over API
- [x] post-merge/deploy test of a non multi-action mistral assistant

## Risk

Medium, could break some pre-existing Mistral interactions but unlikely.

## Deploy Plan

- deploy `core`

### CODE REVIEW

**Comment by spolu** [comment]
(streaming working as expected)

File: core/src/providers/mistral.rs [hunk 1]
```diff
@@ -104,13 +105,16 @@ struct MistralChatMessage {
     pub name: Option<String>,
     #[serde(skip_serializing_if = "Option::is_none")]
     pub tool_calls: Option<Vec<MistralToolCall>>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub tool_call_id: Option<String>,
```
Comment (fontanierh) [inline core/src/providers/mistral.rs:109]: Why can it be None ? If none, we'll need a default no ? IIRC, mistral API requires it

File: core/src/providers/mistral.rs [hunk 2]
```diff
@@ -387,30 +400,46 @@ impl MistralAILLM {
             .map(|m| MistralChatMessage::try_from(m))
             .collect::<Result<Vec<_>>>()?;
 
-        // Mistral AI requires a tool call to be followed by a tool message. We therefore inject a
-        // tool call message before the tool message if missing.
+        // Mistral AI requires a tool call assistant message to be followed by a tool message. We
```
Comment (fontanierh) [inline core/src/providers/mistral.rs:403]: Once we phase-out single action assistants (and therefore the old renderConversationForModel), should we remove this logic ?

File: core/src/providers/mistral.rs [hunk 3]
```diff
@@ -723,6 +753,11 @@ impl MistralAILLM {
             body["tools"] = json!(tools);
         }
 
+        // println!(
```
Comment (fontanierh) [inline core/src/providers/mistral.rs:756]: nit

File: core/src/providers/mistral.rs [hunk 4]
```diff
@@ -746,6 +781,8 @@ impl MistralAILLM {
         body.reader().read_to_end(&mut b)?;
         let c: &[u8] = &b;
 
+        // println!("MistralAI response: {}", String::from_utf8_lossy(c));
```
Comment (fontanierh) [inline core/src/providers/mistral.rs:784]: nit

**Review by fontanierh — APPROVED** [review]

File: core/src/providers/mistral.rs (hunk 1 above)
Comment (spolu) [reply core/src/providers/mistral.rs:109]: They want it only if tool message? and for other they don't want to hear about it?

File: core/src/providers/mistral.rs (hunk 2 above)
Comment (spolu) [reply core/src/providers/mistral.rs:403]: We should yes

File: core/src/providers/mistral.rs (hunk 3 above)
Comment (spolu) [reply core/src/providers/mistral.rs:756]: They are very convenient in favor of keeping them

File: core/src/providers/mistral.rs (hunk 4 above)
Comment (spolu) [reply core/src/providers/mistral.rs:784]: They are very convenient in favor of keeping them

**Comment by spolu** [comment]
I have a single action assistant ready in prod to test post deploy

**Comment by spolu** [comment]
(pivoting for now on understanding the error condition of multi actions mistral pre this fix which is now well reported in the UI)

File: core/src/providers/mistral.rs (hunk 1 above)
Comment (fontanierh) [reply core/src/providers/mistral.rs:109]: Yes nvm me I didn't understand that this was used for all kinds of msg originally 👍

File: core/src/providers/mistral.rs (hunk 3 above)
Comment (fontanierh) [reply core/src/providers/mistral.rs:756]: Quite bearish on commented-out debugging code personally, makes the code look messy and unclean, but as you prefer !

File: core/src/providers/mistral.rs [hunk 5]
```diff
@@ -176,10 +180,19 @@ impl TryFrom<&ChatMessage> for MistralChatMessage {
...
+            tool_calls: match cm.function_calls.as_ref() {
+                Some(fc) => Some(
+                    fc.iter()
+                        .map(|f| MistralToolCall::try_from(f))
+                        .collect::<Result<Vec<_>>>()?,
+                ),
                 None => None,
             },
+            tool_call_id: cm.function_call_id.clone(),
+            // tool_calls: match cm.function_calls.as_ref() {
```
Comment (flvndvd) [inline core/src/providers/mistral.rs:192]: ✂️

File: core/src/providers/mistral.rs [hunk 6]
```diff
@@ -387,30 +400,46 @@ impl MistralAILLM {
...
                         acc.push(MistralChatMessage {
                             role: MistralChatMessageRole::Assistant,
                             name: None,
                             content: None,
                             tool_calls: Some(vec![MistralToolCall {
+                                // If we have a tool_call in the agent message we use it, otherwise
+                                // we generate a new fake one following Mistral format (9 chars, we
+                                // use an hex representation where Mistral uses any char, but
+                                // that's fine).
+                                id: cm
```
Comment (flvndvd) [inline core/src/providers/mistral.rs:430]: That's for the case where no id is provided through the API?

**Review by flvndvd — APPROVED** [review]

File: core/src/providers/mistral.rs (hunk 6 above)
Comment (spolu) [reply core/src/providers/mistral.rs:430]: In v0 assistant we just create a fake function message with no ID I think so we want to fake one here.

File: core/src/providers/mistral.rs (hunk 3 above)
Comment (spolu) [reply core/src/providers/mistral.rs:756]: kk will remove

**Comment by spolu** [comment]
Also aligns function_call event implementations. Checked locally that streamed event are aligned when switching models.

**Review by fontanierh — APPROVED** [review]
Thanks !

**Pull request merged**